lxd/device/proxy.go

```diff
@@ -0,0 +1,343 @@
+		// Ensure that the provided IP address matches the container's IP
+		// address otherwise we could mess with other containers.
+		if ip != "" && IPv4Addr == "" && (address == ip || address == "0.0.0.0") {
+			IPv4Addr = ip
+		}
+
+		ip = devConfig["ipv6.address"]
+		if ip != "" && IPv6Addr == "" && (address == ip || address == "::") {
+			IPv6Addr = ip
+		}
+	}
+
+	if IPv4Addr == "" && IPv6Addr == "" {
+		return fmt.Errorf("NIC IP doesn't match proxy target IP")
+	}
+
+	iptablesComment := fmt.Sprintf("%s (%s)", d.instance.Name(), d.name)
+
+	revert := true
+	defer func() {
+		if revert {
+			if IPv4Addr != "" {
+				iptables.ContainerClear("ipv4", iptablesComment, "nat")
+			}
+
+			if IPv6Addr != "" {
+				iptables.ContainerClear("ipv6", iptablesComment, "nat")
+			}
+		}
+	}()
+
+	for i, lAddr := range listenAddr.Addr {
+		address, port, err := net.SplitHostPort(lAddr)
+		if err != nil {
+			return err
+		}
+		var cPort string
+		if len(connectAddr.Addr) == 1 {
+			_, cPort, _ = net.SplitHostPort(connectAddr.Addr[0])
+		} else {
+			_, cPort, _ = net.SplitHostPort(connectAddr.Addr[i])
+		}
+
+		if IPv4Addr != "" {
+			// outbound <-> container
+			err := iptables.ContainerPrepend("ipv4", iptablesComment, "nat",
+				"PREROUTING", "-p", listenAddr.ConnType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("%s:%s", IPv4Addr, cPort))
+			if err != nil {
+				return err
+			}
+
+			// host <-> container
+			err = iptables.ContainerPrepend("ipv4", iptablesComment, "nat",
+				"OUTPUT", "-p", listenAddr.ConnType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("%s:%s", IPv4Addr, cPort))
+			if err != nil {
+				return err
+			}
+		}
+
+		if IPv6Addr != "" {
+			// outbound <-> container
+			err := iptables.ContainerPrepend("ipv6", iptablesComment, "nat",
+				"PREROUTING", "-p", listenAddr.ConnType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("[%s]:%s", IPv6Addr, cPort))
+			if err != nil {
+				return err
+			}
+
+			// host <-> container
+			err = iptables.ContainerPrepend("ipv6", iptablesComment, "nat",
+				"OUTPUT", "-p", listenAddr.ConnType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("[%s]:%s", IPv6Addr, cPort))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	revert = false
+	return nil
+}
+
+func rewriteHostAddr(addr string) string {
+	proto, addr := networkParseAddr(addr)
+	if proto == "unix" && !strings.HasPrefix(addr, "@") {
+		// Unix non-abstract sockets need to be addressed to the host
+		// filesystem, not be scoped inside the LXD snap.
+		addr = shared.HostPath(addr)
+	}
+	return fmt.Sprintf("%s:%s", proto, addr)
+}
+
+func (d *proxy) setupProxyProcInfo() (*proxyProcInfo, error) {
+	cname := project.Prefix(d.instance.Project(), d.instance.Name())
+	cc, err := lxc.NewContainer(cname, d.state.OS.LxcPath)
```

Comment:
@stgraber the proxy device needs the PID of the running container, I believe so it can enter into the namespace to setup the proxy listeners. Rather than adding something to the device interface that is interface specific, I have used go-lxc directly. Is this OK to do?

Reply:
So from what I remember of our earlier discussions, we want to use `lxc.hook.start-host` to spawn the proxy. Do we get something in the environment there which would let us attach to those namespaces without needing to go grab the pid with go-lxc?

Reply:
@stgraber as discussed, I've added a `postStartHook` slice of functions in place of `restartProxyDevices` function calls.